File: lib/search-utils.ts
const normalizeSearchValue = (value?: string | null) =>
  (value || '')
    .normalize('NFD')
    .replace(/[\u0300-\u036f]/g, '')
    .toLowerCase()
    .trim()
    .replace(/\s+/g, ' ');

export const matchesSearchQuery = (query: string, ...fields: Array<string | null | undefined>) => {
  const normalizedQuery = normalizeSearchValue(query);
  if (!normalizedQuery) return true;
  const haystack = normalizeSearchValue(fields.filter(Boolean).join(' '));
  if (!haystack) return false;
  return normalizedQuery.split(' ').every((term) => haystack.includes(term));
};

File: backend/src/main/java/com/notiflow/util/SearchUtils.java
package com.notiflow.util;

import java.text.Normalizer;
import java.util.Arrays;
import java.util.Locale;
import java.util.Objects;
import java.util.stream.Collectors;

public final class SearchUtils {

    private SearchUtils() {
    }

    public static boolean matchesQuery(String query, String... fields) {
        String normalizedQuery = normalize(query);
        if (normalizedQuery.isBlank()) {
            return true;
        }
        String haystack = normalize(Arrays.stream(fields)
                .filter(Objects::nonNull)
                .collect(Collectors.joining(" ")));
        if (haystack.isBlank()) {
            return false;
        }
        return Arrays.stream(normalizedQuery.split(" "))
                .filter(term -> !term.isBlank())
                .allMatch(haystack::contains);
    }

    public static String normalize(String value) {
        if (value == null) {
            return "";
        }
        String normalized = Normalizer.normalize(value, Normalizer.Form.NFD)
                .replaceAll("\\p{M}+", "")
                .toLowerCase(Locale.ROOT)
                .trim();
        return normalized.replaceAll("\\s+", " ");
    }
}
